torch/csrc/jit/codegen/onednn/interface.cpp

```diff
@@ -0,0 +1,145 @@
+    GRAPH_DUMP(
+        "Before RemoveProfileNodesAndSpecializeTypes. Beginning of LLGA "
+        "optimization pass",
+        g);
+    RemoveProfileNodesAndSpecializeTypes(g);
+    GRAPH_DUMP(
+        "After RemoveProfileNodesAndSpecializeTypes. Before mutation removal",
+        g);
+
+    RemoveTensorMutation(g);
+    RemoveListMutation(g);
+    GRAPH_DUMP("After mutation removal. Before DecomposeOps", g);
+    DecomposeOps(g);
+    GRAPH_DUMP("After DecomposeOps. Before PrepareBinaryForLLGA", g);
+    PrepareBinaryForLLGA(g);
+    GRAPH_DUMP("After PrepareBinaryForLLGA. Before DeferSizeCheck", g);
+    DeferSizeCheck(g);
+    GRAPH_DUMP("After DeferSizeCheck. Before CreateLlgaSubgraphs", g);
+    CreateLlgaSubgraphs(g);
+    GRAPH_DUMP("After CreateLlgaSubgraphs. Before PropagateLayout", g);
+    PropagateLayout(g);
+    GRAPH_DUMP(
+        "After PropagateLayout. Before prepareFusionGroupAndGuardOutputs", g);
+
+    // Add shape guard for profiling mode and wipe the tensor type information
+    // from the IR
+    prepareFusionGroupAndGuardOutputs(g->block());
+    GRAPH_DUMP(
+        "After prepareFusionGroupAndGuardOutputs. Before "
+        "RemoveTensorTypeSpecializations",
+        g);
+    RemoveTensorTypeSpecializations(g);
+    GRAPH_DUMP(
+        "After RemoveTensorTypeSpecializations. End of LLGA optimization pass",
+        g);
+  }
+}
+
+} // namespace onednn
+} // namespace fuser
+
+Operation createLlgaKernel(const Node* node) {
+  auto kernel = std::make_shared<fuser::onednn::LlgaKernel>(node);
+  return [kernel](Stack* stack) {
+    RECORD_FUNCTION(kernel->debugName(), std::vector<c10::IValue>());
+    kernel->run(*stack);
+    return 0;
+  };
+}
+
+RegisterOperators oneDNNFusionGroupOp({
+    torch::jit::Operator(
+        prim::oneDNNFusionGroup,
+        createLlgaKernel,
+        AliasAnalysisKind::INTERNAL_SPECIAL_CASE),
+});
+
+Operation createLlgaGuardKernel(const Node* node) {
+  return [node](Stack* stack) {
+    GRAPH_DEBUG("Guarding node: ", node->kind().toQualString());
+    std::vector<TypePtr> types = node->tys(attr::types);
+    const auto num_inputs = types.size();
+
+    GRAPH_DEBUG("num_inputs to guard: ", num_inputs);
+
+    for (size_t i = 0; i < num_inputs; i++) {
+      GRAPH_DEBUG("checking input ", i);
+      auto& input = peek(stack, i, num_inputs);
+      const c10::TensorTypePtr& guard_tensor_type =
+          types[i]->cast<TensorType>();
+
+      if (!input.isTensor()) {
+        GRAPH_DEBUG("input ", i, " is not a tensor, return false");
+        push(stack, IValue(false));
+        return;
+      }
+      const at::Tensor& tensor = input.toTensor();
+
+      // If input tensor is of mkldnn, it's originated from an upstream
+      // LLGA partition that has passed the check on input shapes.
+      // It is valid to continue here as long as the output shapes from
+      // oneDNN graph partitions are determined by the input shapes.
+      if (tensor.is_mkldnn()) {
```

Comment:
It seems a little suspect to not check sizes here but i don't know the logic well enough.

Reply:
Tensors between two oneDNN Graph partitions can have different (but optimal) layouts (eg. blocked or channels-last), so they're opaque to PyTorch. So, we need not check their shape.

Reply:
We wrap an output tensor of a oneDNN graph partition as MKLDNN tensors only if it's an input to another oneDNN partition, but not a PyTorch op that isn't present in any oneDNN graph partition. Their size & dtype would be same as that expected, but their strides may be different (depending on the layout). Currently, the device can only be CPU. We're overriding guard checks for such tensors based on the assumption that size/dtype/device would be correct, as oneDNN Graph would've generated these tensors for use by a oneDNN Graph partition, as oneDNN Graph is already cognizant of the characteristics of the input tensors of the next partition.